collect_app/src/main/java/org/odk/collect/android/preferences/PreferencesRepository.kt

```diff
@@ -0,0 +1,48 @@
+package org.odk.collect.android.preferences
+
+import android.content.Context
+import androidx.preference.PreferenceManager
+
+class PreferencesRepository(private val context: Context) {
+    @JvmOverloads
+    fun getMetaPreferences(projectId: String = ""): PreferencesDataSource {
+        val preferenceId = META_PREFS_NAME + projectId
+
+        if (!preferences.containsKey(preferenceId)) {
+            preferences[META_PREFS_NAME + projectId] = PreferencesDataSource(context.getSharedPreferences(preferenceId, Context.MODE_PRIVATE))
+        }
+        return preferences[preferenceId]!!
+    }
+
+    @JvmOverloads
+    fun getGeneralPreferences(projectId: String = ""): PreferencesDataSource {
+        val preferenceId = GENERAL_PREFS_NAME + projectId
+
+        if (!preferences.containsKey(preferenceId)) {
+            if (projectId.isBlank()) {
+                preferences[preferenceId] = PreferencesDataSource(PreferenceManager.getDefaultSharedPreferences(context), GeneralKeys.DEFAULTS)
+            } else {
+                preferences[preferenceId] = PreferencesDataSource(context.getSharedPreferences(preferenceId, Context.MODE_PRIVATE), GeneralKeys.DEFAULTS)
+            }
+        }
+        return preferences[preferenceId]!!
+    }
+
+    @JvmOverloads
+    fun getAdminPreferences(projectId: String = ""): PreferencesDataSource {
+        val preferenceId = ADMIN_PREFS_NAME + projectId
+
+        if (!preferences.containsKey(preferenceId)) {
+            preferences[preferenceId] = PreferencesDataSource(context.getSharedPreferences(preferenceId, Context.MODE_PRIVATE), AdminKeys.getDefaults())
+        }
+        return preferences[preferenceId]!!
+    }
+
+    companion object {
+        private val preferences = mutableMapOf<String, PreferencesDataSource>()
```

Comment:
Right. Fixed.

Reply:
I added it back because I had some failing tests... yes it's true that PreferencesDataSourceProvider and it's used like that in 99% of cases but there are 2 or 3 places where a new object is created. I should review those places maybe I can improve it.